### DISCUSSION TITLE
Gfortran compiled on x86, how to compile with clang on openEuler, do I need to modify the source code somewhere?

### DISCUSSION THREAD

**Discussion by 123xyl** [openmopac/mopac#105] [Q&A]
Gfortran compiled on x86, how to compile with clang on openEuler, do I need to modify the source code somewhere?

**Comment by godotalgorithm** [comment]
clang is a C/C++ compiler. MOPAC is written in Fortran, and it needs a Fortran compiler. The Fortran companion to clang is flang [https://github.com/flang-compiler/flang], but I haven't tried to use it and I'm not sure if it is production-ready yet. If you are on an ARM processor, there should be ARM ports of gcc-based compilers that should have a useable version of gfortran or f95.

In principle, MOPAC's CMake build system should work in any computing environment where you have access to (either precompiled or compiled yourself): (1) a Fortran compiler, (2) BLAS and LAPACK libraries, and (3) CMake. OpenEuler is a Linux distribution, but I don't know what package management systems it uses and I don't know what changes it might have made relative to more popular Linux distributions that might confuse the automation of CMake.

The simplest customizations of a CMake build can be made through command-line arguments that are shared by most programs that you might build using CMake. For example, the `-DCMAKE_Fortran_COMPILER=path_to_fortran_compiler` argument sets the compiler if CMake doesn't automatically detect the one you want. MOPAC uses CMake's find_package system for automatically detecting BLAS and LAPACK, which can often fail. As I noted in an earlier Discussion post, the `-DAUTOBLAS=OFF` argument turns off the automated BLAS and LAPACK detection, and  `-DMOPAC_LINK="fortran_link_line"` will set the link line that the compiler uses, which would include `-L` commands to set the paths to libraries and `-l` commands that name specific libraries to link.

MOPAC shouldn't be any harder to build with CMake than most other programs, so you should familiarize yourself with the use of CMake on whatever system you are using.

There is a pre-compiled ARM-based Linux version of MOPAC available on the conda-forge channel of the conda package manager [ linux-aarch64 @ https://anaconda.org/conda-forge/mopac], which might be compatible with an ARM processor running openEuler.

One last warning: someone recently reported that the Mac ARM version of MOPAC on conda-forge is randomly crashing, and whatever is causing this might cause MOPAC to crash on ARM processors with other operating systems. MOPAC plans to fully support ARM in the future, but we presently don't have any conveniently available ARM-based development systems
